File: src/task.cpp
#include "project11_navigation/task.h"
#include <project11_navigation/utilities.h>
#include <ros/ros.h>

#include <project11_navigation/context.h>


namespace project11_navigation
{
Task::Task():
  children_(this),
  last_update_time_(ros::Time::now())
{
}

std::shared_ptr<Task> Task::create(const project11_nav_msgs::TaskInformation& task_msg, TaskPtr parent_task)
{
  std::shared_ptr<Task> new_task;

  if(!new_task)
    new_task = std::make_shared<Task>();
  
  new_task->message_ = task_msg;
  new_task->self_ = new_task;
  new_task->parent_task_ = parent_task;
  return new_task;
}

std::shared_ptr<Task> Task::self()
{
  return self_.lock();
}

bool Task::update(const project11_nav_msgs::TaskInformation& task_msg, bool check_id)
{
  if(!check_id || message_.id == task_msg.id)
  {
    if(message_ != task_msg)
      last_update_time_ = ros::Time::now();
    message_= task_msg;
    return true;
  }
  return false;
}

void Task::update(const std::vector<project11_nav_msgs::TaskInformation>& task_msgs)
{
  for(auto task_msg: task_msgs)
    update(task_msg);
  children_.update(task_msgs);
}

const project11_nav_msgs::TaskInformation & Task::message() const
{
  return message_;
}

void Task::setDone()
{
  message_.done = true;
  last_update_time_ = ros::Time::now();
}

YAML::Node Task::data() const
{
  return YAML::Load(message_.data);
}

void Task::setData(const YAML::Node& data)
{
  std::stringstream ss;
  ss << data;
  message_.data = ss.str();
  last_update_time_ = ros::Time::now();
}

YAML::Node Task::dataItem(std::string key, bool recurse_up) const
{
  auto d = data();
  if(d[key])
    return d[key];
  if(recurse_up && !parent_task_.expired())
    return parent_task_.lock()->dataItem(key, recurse_up);
  return {};
}

YAML::Node Task::status() const
{
  return YAML::Load(message_.status);
}

void Task::setStatus(const YAML::Node& status)
{
  std::stringstream ss;
  ss << status;
  message_.status = ss.str();
  last_update_time_ = ros::Time::now();
}

void Task::setID(std::string id)
{
  message_.id = id;
  last_update_time_ = ros::Time::now();
}

void Task::setChildID(std::shared_ptr<Task> task, std::string id)
{
  task->setID(message_.id + "/" + id);
}

bool Task::done(bool recursive) const
{
  if(recursive)
    return message_.done && children_.allDone(true);
  return message_.done;
}

const TaskList& Task::children() const
{
  return children_;
}

ros::Time Task::lastUpdateTime() const
{
  return last_update_time_;  
}

bool Task::getFirstPose(geometry_msgs::PoseStamped& pose, bool recursive) const
{
  if(!message_.poses.empty())
  {
    pose = message_.poses.front();
    return true;
  }
  if(recursive)
    return children_.getFirstPose(pose, recursive);
  return false;
}

bool Task::getLastPose(geometry_msgs::PoseStamped& pose, bool recursive) const
{
  if(!message_.poses.empty())
  {
    pose = message_.poses.back();
    return true;
  }
  if(recursive)
    return children_.getLastPose(pose, recursive);
  return false;
}

std::shared_ptr<Task> Task::createChildTaskBefore(std::shared_ptr<Task> task, std::string type)
{
  return children_.createTaskBefore(task, type);
}

std::string Task::getChildID(std::string id) const
{
  return message_.id+"/"+id;
}

void Task::clearChildren()
{
  children_.clear();
}


}  // namespace project11_navigation

File: src/project11_navigation/task.py
#!/usr/bin/env python3

from typing import Iterable, Any, List, Tuple, Dict

import rospy
from project11_nav_msgs.msg import TaskInformation

import io
import copy
import yaml

def parentID(task_id: str) -> str:
  if '/' in task_id:
    return task_id.rsplit('/',1)[0]
  return None


class Task:
  '''Help manage TaskInformation messages'''

  def __init__(self, task_information: TaskInformation, task_list: 'TaskList') -> None:
    '''
    Initializes a Task, adding or updating entries in task_list.
    '''
    self.task_information = copy.deepcopy(task_information)
    self.task_list = task_list
    self.last_update_time = rospy.Time.now()
    task_list.tasks[task_information.id] = self
    if not task_information.id in task_list.task_order_ids:
      task_list.task_order_ids.append(task_information.id)
 

  def parentID(self) -> str:
    '''
    Returns the id of the parent task if a child or
    None if this is a top-level task
    '''
    return parentID(self.task_information.id)


  def parent(self) -> 'Task':
    return self.task_list.parentTask(self.task_information.id)


  def children(self) -> List['Task']:
    return self.task_list.childrenTasks(self.task_information.id)

  def update(self, task_info: TaskInformation) -> None:
    '''
    Updates TaskInformation message for this task.
    '''
    if self.task_information.id != task_info.id:
      raise KeyError("Trying to update a task with a different id")
    self.task_information = copy.deepcopy(task_info)
    self.last_update_time = rospy.Time.now()
    
  def sameMessage(self, task_information: TaskInformation) -> bool:
    '''Checks if the incoming TaskInformation message is identical
    to this task's version.
    '''
    this_buffer = io.StringIO()
    self.task_information.serialize(this_buffer)
    other_buffer = io.StringIO()
    task_information.serialize(other_buffer)
    return this_buffer.getvalue() == other_buffer.getvalue()
  
  def message(self) -> TaskInformation:
    '''Returns the TaskInformation message'''
    return self.task_information
  
  def descendantMessages(self) -> List[TaskInformation]:
    '''
    Returns a list of TaskInformation messages for all
    the children tasks recursively.
    '''
    messages = []
    for task in self.children():
      messages.append(task.message())
      messages += messages
    return messages

  def done(self, recursive: bool = False) -> bool:
    '''Returns true if task and optionally the children tasks
    are done. Defaults to not checking children.
    '''
    if not recursive:
      return self.task_information.done
    
    return self.task_information.done and self.task_list.allChildrenDone(self.task_information.id)

  def setDone(self) -> None:
    '''Set the message's done flag to True'''
    self.task_information.done = True
    self.last_update_time = rospy.Time.now()

  def getFullChildID(self, child_local_id: str) -> None:
    '''Generates a full id for a child task of this task.
    The full id is built by adding the id parameter to this task's id and a separating '/'.
    '''
    return self.task_information.id+'/'+child_local_id

  def data(self) -> Any:
    '''Return's the TaskInformation's data decoded from YAML'''
    return yaml.safe_load(self.task_information.data)

  def setData(self, data: Any) -> None:
    '''YAML encodes data and sets it'''
    self.task_information.data = yaml.safe_dump(data)
    self.last_update_time = rospy.Time.now()

  def dataItem(self, key: str, recurse_up: bool = True) -> Any:
    '''
    Returns a data item from the parsed data member of the message
    optionally recursing up the parents.
    '''
    data = self.data()
    if key in data:
      return data[key]
    if recurse_up:
      return self.task_list.parentTask(self.task_information.id).dataItem(key, recurse_up)
    return None

  def status(self) -> Any:
    '''Return's the TaskInformation's status data decoded from YAML'''
    return yaml.safe_load(self.task_information.status)

  def setStatus(self, status: Any) -> None:
    '''YAML encodes status and sets it'''
    self.task_information.status = yaml.safe_dump(status)
    self.last_update_time = rospy.Time.now()

# this is to make type hints work
from project11_navigation.task_list import TaskList
